lib/changeset.ex

```diff
@@ -0,0 +1,162 @@
+           attribute
+         }`"}
+
+  def code_and_description(
+        {"should have at most %{count} item(s)", [count: count, validation: :length, max: _]}
+      ),
+      do: {:invalid_value, "This property should have at most #{count} items."}
+
+  def code_and_description({"must be empty if property given", [property: property]}),
+    do: {:invalid_value, "This property must be empty if #{property} is given."}
+
+  def code_and_description({"time range incomplete", _}),
+    do: {:time_range_incomplete, "Both time range fields must be non-null."}
+
+  def code_and_description({"must be in the future", _}),
+    do: {:future_value_required, "Must be in the future."}
+
+  def code_and_description({"invalid start time", _}),
+    do: {:invalid_start_time, "The start of the time range must be before the end."}
+
+  def code_and_description({"invalid end time", _}),
+    do: {:invalid_end_time, "The end of the time range must be after the start."}
+
+  def code_and_description({"phone or email required", _}),
+    do: {:phone_or_email_required, "Either phone or email must be provided."}
+
+  def code_and_description({"no slot available on device", _}),
+    do: {:no_slot_available, "No slot available on the device."}
+
+  def code_and_description(
+        {"Permanent codes cannot have a start and end time, or recurring fields", _}
+      ),
+      do:
+        {:invalid_activation_type_fields,
+         "Permanent codes cannot have a start and end time, or recurring fields"}
+
+  def code_and_description(
+        {"Recurring codes cannot have a temporary start at or temporary end at time", _}
+      ),
+      do:
+        {:invalid_activation_type_fields,
+         "Recurring codes cannot have a temporary start at or temporary end at time"}
+
+  def code_and_description(
+        {"Each value must be a capitalized day of the week", [invalid_values: invalid_values]}
+      ),
+      do:
+        {:invalid_values,
+         "Each value must be a capitalized day of the week. Invalid values: #{
+           Enum.join(invalid_values, ", ")
+         }"}
+
+  def code_and_description({"Icon not supported", _}),
+    do: {:invalid_icon, "Icon not supported. Please use a valid icon name."}
+
+  def code_and_description(
+        {"should be at most %{count} character(s)",
+         [count: count, validation: :length, kind: :max, type: :string]}
+      ),
+      do: {:invalid_value, "This property should have at most #{count} characters."}
+
+  # V1 of mobile apps respond to this error
+  def code_and_description({"duplicate code", _}),
+    do: {:duplicate_code, "Code already in use."}
+
+  # If a custom validation error was created with an error_code attached,
+  # use that error_code (and the error message as the description)
+  def code_and_description({description, [error_code: error_code]}),
+    do: {error_code, description}
```

Comment:
Where do we draw the line between what goes in this package and what doesn't? Do we have to add something here every time we add custom validation in any project? What happens if we forget?

Reply:
That's a really good point. I think we should go over this in the API Squad meeting on Monday.